<launch>

    <!-- mono or stereo and what ros bag to play -->
    <arg name="max_cameras" default="2" />
    <arg name="use_stereo"  default="true" />

    <!-- saving trajectory path and timing information -->
    <arg name="dosave"      default="false" />
    <arg name="dotime"      default="false" />
    <arg name="path_est"    default="/home/hoangqc/Datasets/Airsim-ros/traj_estimate.txt" />
    <arg name="path_time"   default="/home/hoangqc/Datasets/Airsim-ros/traj_timing.txt" />

    <!-- MASTER NODE! -->
    <node name="run_serial_msckf" pkg="ov_msckf" type="run_serial_msckf" output="screen" clear_params="true" required="true">
<!--    <node name="run_serial_msckf" pkg="ov_msckf" type="run_serial_msckf" output="screen" clear_params="true" required="true" launch-prefix="gdb -ex run &#45;&#45;args">-->

        <!-- bag topics -->
        <param name="topic_imu"      type="string" value="/imu0" />
        <param name="topic_camera0"  type="string" value="/cam0/image_raw" />
        <param name="topic_camera1"  type="string" value="/cam1/image_raw" />
        <rosparam param="stereo_pairs">[0, 1]</rosparam>

        <!-- bag parameters -->
        <param name="path_bag"    type="string" value="/home/hoangqc/Datasets/Airsim-ros/NH-base-20fps.bag" />
        <param name="path_gt"     type="string" value="/home/hoangqc/Datasets/Airsim-ros/1-gt/NH-base-20fps.csv" />
        <param name="bag_start"   type="int"    value="0" />
        <param name="bag_durr"    type="int"    value="-1" />

        <!-- world/filter parameters -->
        <param name="use_fej"                type="bool"   value="true" />
        <param name="use_imuavg"             type="bool"   value="true" />
        <param name="use_rk4int"             type="bool"   value="true" />
        <param name="use_stereo"             type="bool"   value="$(arg use_stereo)" />
        <param name="calib_cam_extrinsics"   type="bool"   value="false" />
        <param name="calib_cam_intrinsics"   type="bool"   value="false" />
        <param name="calib_cam_timeoffset"   type="bool"   value="false" />
        <param name="calib_camimu_dt"        type="double" value="0.00" />
        <param name="max_clones"             type="int"    value="11" />
        <param name="max_slam"               type="int"    value="75" />
        <param name="max_slam_in_update"     type="int"    value="25" /> <!-- 15 seems to work well -->
        <param name="max_msckf_in_update"    type="int"    value="40" />
        <param name="max_cameras"            type="int"    value="$(arg max_cameras)" />
        <param name="dt_slam_delay"          type="double" value="3" />
        <param name="init_window_time"       type="double" value="0.75" />
        <param name="init_imu_thresh"        type="double" value="1.5" />
        <rosparam param="gravity">[0.0,0.0,9.80665]</rosparam>
        <param name="feat_rep_msckf"         type="string" value="GLOBAL_3D" />
        <param name="feat_rep_slam"          type="string" value="ANCHORED_FULL_INVERSE_DEPTH" />
        <param name="feat_rep_aruco"         type="string" value="ANCHORED_FULL_INVERSE_DEPTH" />

        <!-- zero velocity update parameters -->
        <param name="try_zupt"               type="bool"   value="false" />
        <param name="zupt_chi2_multipler"    type="int"    value="1" />
        <param name="zupt_max_velocity"      type="double" value="0.5" />
        <param name="zupt_noise_multiplier"  type="double" value="25" />

        <!-- timing statistics recording -->
        <param name="record_timing_information"   type="bool"   value="$(arg dotime)" />
        <param name="record_timing_filepath"      type="string" value="$(arg path_time)" />

        <!-- tracker/extractor properties -->
        <param name="use_klt"            type="bool"   value="true" />
        <param name="num_pts"            type="int"    value="250" />
        <param name="fast_threshold"     type="int"    value="15" />
        <param name="grid_x"             type="int"    value="9" />
        <param name="grid_y"             type="int"    value="7" />
        <param name="min_px_dist"        type="int"    value="5" />
        <param name="knn_ratio"          type="double" value="0.65" />
        <param name="downsample_cameras" type="bool"   value="false" />
        <param name="multi_threading"    type="bool"   value="true" />

        <param name="fi_max_dist"        type="double" value="100" />
        <param name="fi_max_baseline"    type="double" value="999" />
        <param name="fi_max_cond_number" type="double" value="18000" />

        <!-- aruco tag/mapping properties -->
        <param name="use_aruco"        type="bool"   value="false" />
        <param name="num_aruco"        type="int"    value="1024" />
        <param name="downsize_aruco"   type="bool"   value="true" />

        <!-- sensor noise values / update -->
        <param name="up_msckf_sigma_px"            type="double"   value="1" />
        <param name="up_msckf_chi2_multipler"      type="double"   value="1" />
        <param name="up_slam_sigma_px"             type="double"   value="1" />
        <param name="up_slam_chi2_multipler"       type="double"   value="1" />
        <param name="up_aruco_sigma_px"            type="double"   value="1" />
        <param name="up_aruco_chi2_multipler"      type="double"   value="1" />
        <param name="gyroscope_noise_density"      type="double"   value="8.73e-5" />
        <param name="gyroscope_random_walk"        type="double"   value="9.97e-7" />
        <param name="accelerometer_noise_density"  type="double"   value="2.35e-3" />
        <param name="accelerometer_random_walk"    type="double"   value="1.25e-5" />


        <!-- camera intrinsics -->
        <rosparam param="cam0_wh">[640, 400]</rosparam>
        <rosparam param="cam1_wh">[640, 400]</rosparam>
        <param name="cam0_is_fisheye" type="bool" value="false" />
        <param name="cam1_is_fisheye" type="bool" value="false" />
        <rosparam param="cam0_k">[170.147, 170.147, 320.0, 200.0]</rosparam>
        <rosparam param="cam0_d">[0.0, 0.0, 0.0, 0.0]</rosparam>
        <rosparam param="cam1_k">[170.147, 170.147, 320.0, 200.0]</rosparam>
        <rosparam param="cam1_d">[0.0, 0.0, 0.0, 0.0]</rosparam>


        <!-- camera extrinsics -->
        <rosparam param="T_C0toI">
            [
            0.0, 0.0, 1.0, 0.0,
            1.0, 0.0, 0.0, -0.15,
            0.0, 1.0, 0.0, 0.0,
            0.0, 0.0, 0.0, 1.0
            ]
        </rosparam>
        <rosparam param="T_C1toI">
            [
            0.0, 0.0, 1.0, 0.0,
            1.0, 0.0, 0.0, 0.15,
            0.0, 1.0, 0.0, 0.0,
            0.0, 0.0, 0.0, 1.0
            ]
        </rosparam>

    </node>

    <!-- record the trajectory if enabled -->
    <group if="$(arg dosave)">
        <node name="recorder_estimate" pkg="ov_eval" type="pose_to_file" output="screen" required="true">
            <param name="topic"      type="str" value="/ov_msckf/poseimu" />
            <param name="topic_type" type="str" value="PoseWithCovarianceStamped" />
            <param name="output"     type="str" value="$(arg path_est)" />
        </node>
    </group>


</launch>